### PR TITLE
Add support for atomic instructions

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds support for the currently proposed atomic instructions in the
wasm spec. The implementaiton matches what LLVM currently has implemented, and
while this may be tweaked over time it should be good enough to work with LLVM
today!

https://github.com/WebAssembly/threads/blob/master/proposals/threads/Overview.md

### CODE REVIEW

File: src/elements/module.rs [hunk 1]
```diff
@@ -702,8 +702,7 @@ mod integration_tests {
 	}
 
 	#[test]
-	#[should_panic]
-	fn wrong_varuint1_case() {
+	fn varuint1_case() {
```
Comment (pepyakin) [inline src/elements/module.rs]: Why this change was made?

**Comment by NikVolf** [comment]
It's not strictly WebAssembly v1, I wonder if people might want that only v1 features should be loaded

Let me think about it a bit

**Comment by pepyakin** [comment]
@NikVolf I've had the same concern and I've concluded that this is ok for parity-wasm and if one doesn't want to use non-v1 features then they will need to handle it appropriately. 

For example, we would need to update gas-metering rules in wasm-utils to include atomic ops (e.g. by assigning them a separate category in the `rules` and setting it to forbidden). In wasmi we would be able to reject modules that contain any traces threads proposal at validation time (unless the proposal is implemented and enabled via some sort of flag similar to how `wabt` handing this right now).

File: src/elements/module.rs (hunk 1 above)
Comment (alexcrichton) [reply src/elements/module.rs]: I was curious too! Apparently though this is a valid module with shared memory which fails validation earlier but passes afterwards

**Comment by NikVolf** [comment]
Thanks @alexcrichton )
